### DISCUSSION TITLE
Threading videostream-cli

### DISCUSSION THREAD

**Discussion by jsimdb** [aiortc/aiortc#611] [Q&A]
Hi there, 

I'm trying to run two clients in a single script using multithreading. The purpose of this is for automated testing of a WebRTC connection. I want to input a video file, and get a recorded video file as an output.

My script is based off the [videostream-cli ](https://github.com/aiortc/aiortc/tree/main/examples/videostream-cli)example. I have two threads that essentially run this example, one as an offer, another as an answer. To communicate between the threads, I'm trying to modify the [CopyAndPasteSignalling](https://github.com/aiortc/aiortc/blob/main/src/aiortc/contrib/signaling.py) class to use a Queue.

The [`run()` ](https://github.com/aiortc/aiortc/blob/main/examples/videostream-cli/cli.py#L80) function from videostream-cli is encapsulated into the class `client`

```python
from aiortc import RTCPeerConnection
from aiortc.contrib.media import MediaPlayer, MediaRecorder, MediaBlackhole
from aiortc.contrib.signaling import add_signaling_arguments, CopyAndPasteSignaling

from clients import Connection, Media, Client
from custom_signalling import QueueSignaling

from threading import Thread
from queue import Queue

def create_client(send_queue, receive_queue, role, play_from=None, record_to=None, client_id=None):
    # create signaling and peer connection
    signaling = QueueSignaling(receive_queue, send_queue, role)
    pc = RTCPeerConnection()
    connection = Connection(pc, signaling)

    # create media source
    player = MediaPlayer(play_from) if play_from else None
    recorder = MediaRecorder(record_to) if record_to else MediaBlackhole()
    media = Media(player, recorder)

    return Client(connection, media, role, client_id=client_id)

def client_callback(client):
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    try:
        loop.run_until_complete(client.run())
    except KeyboardInterrupt:
        pass
    finally:
        # cleanup
        loop.run_until_complete(client.close())


def main(play_from, record_to):
    q1 = Queue()
    q2 = Queue()
    client_1 = create_client(q1, q2, "offer", play_from=play_from, client_id=1)
    client_2 = create_client(q2, q1, "answer", record_to=record_to, client_id=2)

    c1_thread = Thread(target=client_callback, args=(client_1,))
    c2_thread = Thread(target=client_callback, args=(client_2,))

    c1_thread.start()
    c2_thread.start()


if __name__ == "__main__":

    args = parse_args()

    if args.verbose:
        logging.basicConfig(level=logging.DEBUG)

    main(args.play_from, args.record_to)
```

I can see that messages are successfully being sent and received However, the video isn't being recorded. I've determined this is because the player on the offer client isn't sending anything. The answer client is able to record video if the offer is from a non-threaded client. Therefore I think threading the offer has caused an issue somewhere. 

Any ideas on why threading has broken video playback?

**Comment by jlaine** [comment]
Why do you absolutely need to use threading, couldn't you use two processes and a unix pipe?
